src/irr/asset/CImageLoaderOpenEXR.cpp

```diff
@@ -41,17 +41,69 @@ namespace irr
 
 			SContext ctx;
 
-			readVersionField(_file, ctx);
-			readHeader(fileName, ctx);
+			// readVersionField(_file, ctx);
+			// readHeader(fileName, ctx);
 
 			Array2D<Rgba> pixels;
 			int width;
 			int height;
 
 			readRgba(fileName, pixels, width, height);
 
-			// TODO put data into buffer, consider OpenEXR 2.0 supported saving layouts 
-			return SAssetBundle{};
+			constexpr uint32_t MAX_PITCH_ALIGNMENT = 8u;									   // OpenGL cannot transfer rows with arbitrary padding
+			auto calcPitchInBlocks = [](uint32_t width, uint32_t blockByteSize) -> uint32_t	   // try with largest alignment first
+			{
+				auto rowByteSize = width * blockByteSize;
+				for (uint32_t _alignment = MAX_PITCH_ALIGNMENT; _alignment > 1u; _alignment >>= 1u)
+				{
+					auto paddedSize = core::alignUp(rowByteSize, _alignment);
+					if (paddedSize % blockByteSize)
+						continue;
+					return paddedSize / blockByteSize;
+				}
+				return width;
+			};
+
+			ICPUImage::SCreationParams params;
+			params.format = EF_R16G16B16A16_SFLOAT;
+			params.type = ICPUImage::ET_2D;;
+			params.flags = static_cast<ICPUImage::E_CREATE_FLAGS>(0u);
+			params.samples = ICPUImage::ESCF_1_BIT;
+			params.extent.width = width;
+			params.extent.height = height;
+			params.extent.depth = 1u;
+			params.mipLevels = 1u;
+			params.arrayLayers = 1u;
+
+			auto& image = ICPUImage::create(std::move(params));
+
+			auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(image->getImageDataSizeInBytes());
+			auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1u);
+			ICPUImage::SBufferCopy& region = regions->front();
+			//region.imageSubresource.aspectMask = ...; // waits for Vulkan
+			region.imageSubresource.mipLevel = 0u;
+			region.imageSubresource.baseArrayLayer = 0u;
+			region.imageSubresource.layerCount = 1u;
+			region.bufferOffset = 0u;
+			region.bufferRowLength = calcPitchInBlocks(width, getTexelOrBlockBytesize(image->getCreationParameters().format));
```

Comment:
you need to take alignment into account as well, given by calcPitchInBlocks ()

Think of it as a fake/padded width.

Reply:
isn't it good? (outdated comment?)

Reply:
calcPitchInBlocks doesn't take alignment into account

Reply:
well, I checked it and this is a copy of PNG loader `calcPitchInBlocks`
what did I miss?

Reply:
well, you mean alpha channel right?

Reply:
pitchInBlocks is a fake "length", as in while the image can be a 200x200 you can make the length 252.

This only affects the buffer unpack to image when making a GL texture from a GL Buffer.

This means that the following equation to compute the output address is wrong
https://github.com/buildaworldnet/IrrlichtBAW/pull/403/files#diff-63315eefc23fcfd4f83552a09eee0c58R240

This is why PNG Loader uses `getPitch` or something like that instead of the image width, beause of the possible padding at the end of each row when storing image data in a buffer.

Reply:
to fix

Reply:
so assume I'm handling 250th pixel, what should I fill it with?
> pitchInBlocks is a fake "length", as in while the image can be a 200x200 you can make the length 252.

Reply:
I'm filling those fake values with 0

Reply:
dont fill fake with anything, it wont ever be used.

You just need to change your array output addressing code, from `x+y*realWidth` to `x+y*bufferRowLength`

Reply:
great